src/client/clientmap.cpp

```diff
@@ -610,3 +635,190 @@ void ClientMap::PrintInfo(std::ostream &out)
+						drawbufs.add(buf, block_pos, layer);
+				}
+			}
+		}
+	}
+
+	TimeTaker draw("Drawing shadow mesh buffers");
+
+	core::matrix4 m; // Model matrix
+	v3f offset = intToFloat(m_camera_offset, BS);
+
+	// Render all layers in order
+	for (auto &lists : drawbufs.lists) {
+		for (MeshBufList &list : lists) {
+			// Check and abort if the machine is swapping a lot
+			if (draw.getTimerTime() > 1000) {
+				infostream << "ClientMap::renderMapShadows(): Rendering "
+						"took >1s, returning." << std::endl;
+				return;
+			}
+
+			for (auto &pair : list.bufs) {
+				scene::IMeshBuffer *buf = pair.second;
+
+				// override some material properties
+				video::SMaterial local_material = buf->getMaterial();
+				local_material.MaterialType = material.MaterialType;
+				local_material.BackfaceCulling = material.BackfaceCulling;
+				local_material.FrontfaceCulling = material.FrontfaceCulling;
+				local_material.Lighting = false;
+				driver->setMaterial(local_material);
+
+				v3f block_wpos = intToFloat(pair.first * MAP_BLOCKSIZE, BS);
+				m.setTranslation(block_wpos - offset);
+
+				driver->setTransform(video::ETS_WORLD, m);
+				driver->drawMeshBuffer(buf);
+				vertex_count += buf->getVertexCount();
+			}
+
+			drawcall_count += list.bufs.size();
+		}
+	}
+
+	g_profiler->avg(prefix + "draw meshes [ms]", draw.stop(true));
+	g_profiler->avg(prefix + "vertices drawn [#]", vertex_count);
+	g_profiler->avg(prefix + "drawcalls [#]", drawcall_count);
+}
+
+/*
+	Custom update draw list for the pov of shadow light.
+*/
+void ClientMap::updateDrawListShadow(
+		v3f shadow_light_pos, v3f shadow_light_dir, float shadow_range)
+{
+	ScopeProfiler sp(g_profiler, "CM::updateDrawListShadow()", SPT_AVG);
+
+	const v3f camera_position = shadow_light_pos;
+	const v3f camera_direction = shadow_light_dir;
+	// I "fake" fov just to avoid creating a new function to handle orthographic
+	// projection.
+	const f32 camera_fov = m_camera_fov * 1.9f;
+
+	v3s16 cam_pos_nodes = floatToInt(camera_position, BS);
+	v3s16 p_blocks_min;
+	v3s16 p_blocks_max;
+	getBlocksInViewRange(cam_pos_nodes, &p_blocks_min, &p_blocks_max, shadow_range);
+
+	std::vector<v2s16> blocks_in_range;
+
+	for (auto &i : m_drawlist_shadow) {
+		MapBlock *block = i.second;
+		block->refDrop();
+	}
+	m_drawlist_shadow.clear();
+
+	// We need to append the blocks from the camera POV because sometimes
+	// they are not inside the light frustum and it creates glitches.
+	// This could potentially be removed if we figure out why they are missing
+	// from the light frustum.
+	for (auto &i : m_drawlist) {
+		i.second->refGrab();
+		m_drawlist_shadow[i.first] = i.second;
+	}
+
+	// Number of blocks currently loaded by the client
+	u32 blocks_loaded = 0;
+	// Number of blocks with mesh in rendering range
+	u32 blocks_in_range_with_mesh = 0;
+	// Number of blocks occlusion culled
+	u32 blocks_occlusion_culled = 0;
+
+	for (auto &sector_it : m_sectors) {
+			MapSector *sector = sector_it.second;
```

Comment:
all this block indent is too wide

Reply:
fixed, but I'm gonna apply linter at the end of this review .)